### PR TITLE
Numpy2 changes

### PR DESCRIPTION
**Pull request by hyanwong**
With this, almost all tests pass under numpy2, except for one of the 8-bit conversion routines. Do you know what's going on here @benjeffery ?

```python
    def get_site_genotypes_subset(self, site_id, samples):
        start = site_id * self.encoded_genotypes_size
        g = np.zeros(len(samples), dtype=np.int8)
        if self.genotype_encoding == constants.GenotypeEncoding.ONE_BIT:
            for j, u in enumerate(samples):
                byte_index = u // 8
                bit_index = u % 8
                byte = self.genotype_store[start + byte_index]
                mask = 1 << bit_index
                g[j] = int((byte & mask) != 0)
        else:
            for j, u in enumerate(samples):
>               g[j] = self.genotype_store[start + u]
>> E               OverflowError: Python integer 255 out of bounds for int8

```

### CODE REVIEW

**Comment by hyanwong** [comment]
P.s. The np.asarray changes are as a result of this:

```
>       self.data["sites/time"][:] = np.array(value, dtype=np.float64, copy=False)
E       ValueError: Unable to avoid copy while creating an array as requested.
E       If using `np.array(obj, copy=False)` replace it with `np.asarray(obj)` to allow a copy when needed (no behavior change in NumPy 1.x).
E       For more details, see https://numpy.org/devdocs/numpy_2_0_migration_guide.html#adapting-to-changes-in-the-copy-keyword.
```

**Comment by hyanwong** [comment]
Also there is a comment in constants.py "# TODO Change these to use the enum.IntFlag class", but instead I have forced them into the np.uint32 class, so we can compare with the flags array. I'm not sure there's a way to use enum.IntFlag such that the constant values are treated as np.uint32? I think msprime uses enum.IntFlag for some bitflags now, so has it been figured out there, or is this uint+numpy2 thing likely to bite there too?

**Comment by jeromekelleher** [comment]
I think the uint32 thing is more important than making them nice Enums. The msprime enumerations will have the same problems as using standard Python integers.

**Comment by jeromekelleher** [comment]
Looks like we might be assuming uint8 for genotypes currently somewhere, not int8?

**Comment by hyanwong** [comment]
> Looks like we might be assuming uint8 for genotypes currently somewhere, not int8?

Good call. That looks like here: https://github.com/tskit-dev/tsinfer/blob/9d8f9349c953da269d69b1b6a6352f10c7c5abb9/tsinfer/algorithm.py#L90

We can't use np.int8 for the one-bit encoding, because `np.unpackbits` requires `uint`, so I think we'll have to define the `genotype_store` as np.uint8 for the one-bit encoding or np.int8 for the 8-bit one. I'll add that to this PR.

**Comment by hyanwong** [comment]
>  I'll add that to this PR.

Do we need to change anything in the C interface to account for uint vs int genotypes when using 1-bit vs 8-bit packing? I don't think so: it appears as if that just uses a raw list of bytes (see https://github.com/tskit-dev/tsinfer/blob/6b2116de58be2b77a69615cf852b5223f7e4aeea/lib/ancestor_builder.c#L270)

**Comment by benjeffery** [comment]
I'm confused - I thought we needed numpy<2 for numba? Or are you saying that installing >=2 works?

**Comment by hyanwong** [comment]
The most recent numba version is compatible with numpy-2: https://numba.readthedocs.io/en/stable/release/0.60.0-notes.html, but:

> "users should expect some type and numerical deviations with respect to normal Python behavior while using Numba with NumPy 2.0."

It seems to pass the test suite however. I guess we might as well merge this PR in anticipation? Assuming what I've done is sensible.

File: tsinfer/algorithm.py [hunk 1]
```diff
@@ -86,9 +86,11 @@ def __init__(
         self.encoded_genotypes_size = num_samples
         if genotype_encoding == constants.GenotypeEncoding.ONE_BIT:
             self.encoded_genotypes_size = num_samples // 8 + int((num_samples % 8) != 0)
-        self.genotype_store = np.zeros(
-            max_sites * self.encoded_genotypes_size, dtype=np.uint8
-        )
+            self.genotype_store = np.zeros(
+                max_sites * self.encoded_genotypes_size, dtype=np.uint8
+            )
+        else:
```
Comment (jeromekelleher) [inline tsinfer/algorithm.py:92]: I don't think it's a good idea to switch between uint8 and int8 - we should use one or the the other consistently.

I think the simplest thing to do is to stick with uint8 internally for the genotype store, but validate on the way in that values are less than <127.

**Review by benjeffery — APPROVED** [review]

**Review by jeromekelleher — COMMENTED** [review]
We should just check that values are less then 127, there's no point in switching dtypes here.

**Comment by jeromekelleher** [comment]
I assume we're silently masking an error here at the moment by wrapping 255 to -127. Much better to update the test involved to not use 255.

**Comment by hyanwong** [comment]
> I assume we're silently masking an error here at the moment by wrapping 255 to -127. Much better to update the test involved to not use 255.

Aren't we using -1 to mean "missing data" (which we don't allow in the 1-bit encoding)? The failing test is `tests/test_inference.py::TestMissingDataRoundTrip::test_simple_example`

File: tsinfer/algorithm.py (hunk 1 above)
Comment (hyanwong) [reply tsinfer/algorithm.py:92]: If we stick with uint8, we could preserve `-1` values on output by using `.astype(np.int8)`? E.g.

```python
def get_site_genotypes_subset(self, site_id, samples):
        ...
        else:
            for j, u in enumerate(samples):
                # NB missing data (-1) is stored as 255 in the genotype_store
                g[j] = self.genotype_store[start + u].astype(np.int8)

Comment (hyanwong) [reply tsinfer/algorithm.py]: Yeah, we do check that already. The problem is when we are doing the standard 8-bit encoding, using np.uint8 for the store, sticking a "-1" in there (which weirdly we seem to be able to do just fine, and it gets stored as 255), then pulling it out again.

Is the `astype` workaround acceptable?

Comment (hyanwong) [reply tsinfer/algorithm.py]: (in the latest commit I also added an explicit assert for <=127 for the 8-bit encoding)

Comment (jeromekelleher) [reply tsinfer/algorithm.py]: Ah, we're already casting the output to a int8. Yeah, what you have is good then.

**Comment by benjeffery** [comment]
@mergifyio rebase

**Pull request merged**